ezmlm-receipt.1
.\" $Id$
.TH ezmlm-receipt 1
.SH NAME
ezmlm-receipt \- handle sublist replies at the main list
.SH SYNOPSIS
.B ezmlm-receipt
[
.B \-dD
]
.I dir
.SH DESCRIPTION
.B ezmlm-receipt
handles mail from sublists of the main list
stored in
.I dir
and, if it exists, the associated main digest list.
.B ezmlm-receipt
is used for main list with only sublists as subscribers. It is not
useful for normal mailing lists.

.B ezmlm-receipt
is normally invoked from a
.B .qmail
file.
It reads a mail message from its standard input,
and a mail envelope from the
.BR SENDER ,
.BR LOCAL ,
and
.BR HOST
environment variables.

.B ezmlm-receipt
has two functions. It takes the place of
.B ezmlm-receipt(1)
for main lists that have only sublists as subscribers.
Bounces do not lead to removal of the bouncing
address.
.B emzlm-receipt
saves the
bounce, and logs the event to the mail log.
.B ezmlm-receipt
also saves the bounce to
.I dir\fB/bounce/dnnn.ppp
where
.B nnn
is a time stamp and
.B ppp
the PID of the process. A maximum of 50 bounces are saved. Further bounces
are discarded until room is made in the bounce directory by manually removing
files.
If the list uses and SQL
database, the even is logged to the _mlog table instead, and only if the
distribution cookie matches. The bounce is still saved, but only if it
contained a matching distribution cookie.

.B ezmlm-receipt
also handles feedback messages from the sublist. These are used to monitor
that the sublist is able to send out messages. Receipt of feedback messages
is logged to the mail log.
If the list uses and SQL
database, the even is logged to the _mlog table instead, and only if the
distribution cookie matches.
.SH OPTIONS
.TP
.B \-d
.B ezmlm-receipt
will assume the message is for a digest list.
Normally,
.B ezmlm-receipt
will autodetect this from the bounce address. Autodetection makes
.B ezmlm-receipt
less flexible and will be removed in future versions.
.TP
.B \-D
.B ezmlm-receipt
will assume that the bounce is for a normal (non-digest) list.
Normally,
.B ezmlm-receipt
will autodetect this from the bounce address. Autodetection makes
.B ezmlm-receipt
less flexible and will be removed in future versions.
.B \-D
will become the default.
.SH ADDRESSES
.B ezmlm-receipt
handles mail sent to any of the following addresses:
.TP
.I local\fB\-return\-
Trash.
A help message or subscription acknowledgment bounced.
.TP
.I local\fB\-return\-\fImsg\fB\-\fIbox\fB=\fIdomain
Distribution bounce.
Message number
.I msg
was lost on the way to the sublist
.IR box\fB@\fIdomain .
.TP
.I local\fB\-return\-\fImsg\fB\-
Pre-VERP distribution bounce, in QSBMF format.
Message number
.I msg
was lost on the way to one or more sublists;
.B ezmlm-receipt
will parse the bounce to figure out the addresses.
.TP
.I local\fB\-return\-receipt
Feedback message from a sublist.
.B ezmlm-receipt
will figure out message number and sublist address from the envelope sender.

For all the above addresses if,
.I local
is followed by
.IR \-digest ,
messages are assumed to relate to the digest list, and are stored in
.I dir\fB/digest/bounce
rather than in
.I dir \fB/bounce .
.SH "SEE ALSO"
ezmlm-manage(1),
ezmlm-make(1),
ezmlm-return(1),
ezmlm-send(1),
ezmlm-sub(1),
ezmlm-unsub(1),
ezmlm-weed(1),
ezmlm(5),
qmail-command(8)


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

ezmlm-idx-7.1.0, ????-??-??
===========================

- Added support for SQLite3 subscriber databases.
  Thanks to Mike Tedder.

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Added support for decoding sender addresses mangled with a BATV "btv1" tag.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

- Renamed to "setup" target to the more standard "install".

- Modified ezmlm-split to exit silently if there is not "split" file,
  and added it back into the "manager" file to properly handle subscribe
  and unsubscribe requests destined for sublists.

- Made the ezmlm-reject -h (obey headerreject) option the default.

- Fixed ezmlm-send to obey the "addtrailer" setting instead of looking
  for a "text/trailer" file.

- Added support for custom subject lines to ezmlm-get (for digests).

- Fixed handling of adding the trailer on old lists.

- Fixed adding the trailer to posts encoded with base64.

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.

Smaller changes:

- Fixed a bug in ezmlm-request that caused a segfault when attempting to
  execute the "which" command.

- Added new program ezmlm-checksub to replace the
	ezmlm-issubn ... || { echo error; exit 100; }
  lines in the ezmlmrc files.  ezmlm-issubn is now deprecated for use in
  .qmail files, but still exists for backwards compatibility.

- Added ezmlm-weed to the confirmer control files.

- Added <#a#> substitution for the local part of the accept address.

- ezmlm-warn will now process bounces for both the main and digest lists
  if neither the -d nor the -D options are given.

- Added several control files for programs:
	dir/digcount		ezmlm-tstdig
	dir/digestcode		ezmlm-get
	dir/digformat		ezmlm-get
	dir/digsize		ezmlm-tstdig
	dir/digtime		ezmlm-tstdig
	dir/modcanedit		ezmlm-manage
	dir/modcanlist		ezmlm-manage
	dir/modgetonly		ezmlm-get
	dir/modpostonly		ezmlm-store
	dir/nosubconfirm	ezmlm-manage
	dir/nounsubconfirm	ezmlm-manage
	dir/nowarn		ezmlm-warn
	dir/subgetonly		ezmlm-get
	dir/noreturnposts	ezmlm-clean

- Removed the ezmlm-make -4 option in favor of the above files.

- Fixed handling of digest bouncer in ezmlm-dispatch.

- Eliminated the use of vfork for portability.


ezmlm-receipt.c
/*$Id: ezmlm-receipt.c,v 1.10 1999/02/05 04:57:44 lindberg Exp $*/
/*$Name: ezmlm-idx-0324 $*/
/* Handles receipts and bounces from sublists at the main list */
/* Set up instead of ezmlm-return in DIR/bouncer of main list */

#include <sys/types.h>
#include "direntry.h"
#include "stralloc.h"
#include "str.h"
#include "env.h"
#include "slurp.h"
#include "getconf.h"
#include "strerr.h"
#include "byte.h"
#include "case.h"
#include "quote.h"
#include "getln.h"
#include "substdio.h"
#include "error.h"
#include "readwrite.h"
#include "fmt.h"
#include "now.h"
#include "seek.h"
#include "idx.h"
#include "errtxt.h"

#define FATAL "ezmlm-receipt: fatal: "
#define INFO "ezmlm-receipt: info: "

void die_usage()
{
  strerr_die1x(100,"ezmlm-receipt: usage: ezmlm-receipt [-dD] dir");
}

void die_nomem() { strerr_die2x(111,FATAL,ERR_NOMEM); }

void die_badaddr()
{
  strerr_die2x(100,FATAL,ERR_BAD_ADDRESS);
}
void die_trash()
{
  strerr_die2x(0,INFO,"trash address");
}

stralloc line = {0};
stralloc quoted = {0};
stralloc intro = {0};
stralloc bounce = {0};
stralloc header = {0};
stralloc failure = {0};
stralloc paragraph = {0};
stralloc ddir = {0};
stralloc outhost = {0};
stralloc outlocal = {0};
stralloc inlocal = {0};
stralloc tagline = {0};
stralloc listaddr = {0};
stralloc fndate = {0};
stralloc fndir = {0};
stralloc fndatenew = {0};

void die_datenew()
{ strerr_die4sys(111,FATAL,ERR_WRITE,fndatenew.s,": "); }
void die_msgin()
{ strerr_die2sys(111,FATAL,ERR_READ_INPUT); }

char strnum[FMT_ULONG];
char inbuf[1024];
substdio ssin;

char outbuf[256];	/* small - rarely used */
substdio ssout;

unsigned long when;
unsigned long addrno = 0L;

char *sender;
char *dir;
char *workdir;
void **psql = (void **) 0;
stralloc listno = {0};


void doit(addr,msgnum,when,bounce)
/* Just stores address\0nsgnum\0 followed by bounce. File name is          */
/* dttt.ppp[.n], where 'ttt' is a time stamp, 'ppp' the pid, and 'n' the   */
/* number when there are more than 1 addresses in a pre-VERP bounce. In    */
/* this case, the first one is just dttt.ppp, the decond dttt.ppp.2, etc.  */
/* For a main list, bounces come from sublists. They are rare and serious. */
char *addr;
unsigned long msgnum;
unsigned long when;
stralloc *bounce;
{
  int fd;
  unsigned int pos;
  DIR *bouncedir;
  direntry *d;
  unsigned int no;

  if (!stralloc_copys(&fndir,workdir)) die_nomem();
  if (!stralloc_cats(&fndir,"/bounce")) die_nomem();
  if (!stralloc_0(&fndir)) die_nomem();
  bouncedir = opendir(fndir.s);
  if (!bouncedir)
    if (errno != error_noent)
      strerr_die4sys(111,FATAL,ERR_OPEN,line.s,": ");
    else
      strerr_die3x(111,FATAL,fndir.s,ERR_NOEXIST);

  no = MAX_MAIN_BOUNCES;	/* no more than this many allowed */
  while (no && (d = readdir(bouncedir))) {
    if (str_equal(d->d_name,".")) continue;
    if (str_equal(d->d_name,"..")) continue;
    --no;
  }
  closedir(bouncedir);
  if (!no)			/* max no of bounces exceeded */
    strerr_die2x(0,INFO,ERR_MAX_BOUNCE);
				/* save bounce */
  if (!stralloc_copys(&fndate,workdir)) die_nomem();
  if (!stralloc_cats(&fndate,"/bounce/d")) die_nomem();
  pos = fndate.len - 1;
  if (!stralloc_catb(&fndate,strnum,fmt_ulong(strnum,when))) die_nomem();
  if (!stralloc_cats(&fndate,".")) die_nomem();
  if (!stralloc_catb(&fndate,strnum,fmt_ulong(strnum,(unsigned long) getpid())))
	 die_nomem();
  if (addrno) {	/* so that pre-VERP bounces make a d... file per address */
		/* for the first one we use the std-style fname */
    if (!stralloc_cats(&fndate,".")) die_nomem();
    if (!stralloc_catb(&fndate,strnum,fmt_ulong(strnum,addrno))) die_nomem();
  }
  addrno++;	/* get ready for next */
  if (!stralloc_0(&fndate)) die_nomem();
  if (!stralloc_copy(&fndatenew,&fndate)) die_nomem();
  fndatenew.s[pos] = 'D';

  fd = open_trunc(fndatenew.s);
  if (fd == -1) die_datenew();
  substdio_fdbuf(&ssout,write,fd,outbuf,sizeof(outbuf));
  if (substdio_puts(&ssout,addr) == -1) die_datenew();
  if (substdio_put(&ssout,"",1) == -1) die_datenew();
  if (substdio_put(&ssout,strnum,fmt_ulong(strnum,msgnum)) == -1)
	die_datenew();
  if (substdio_put(&ssout,"",1) == -1) die_datenew();

  if (substdio_puts(&ssout,"Return-Path: <") == -1) die_datenew();
  if (!quote2(&quoted,sender)) die_nomem();
  if (substdio_put(&ssout,quoted.s,quoted.len) == -1) die_datenew();
  if (substdio_puts(&ssout,">\n") == -1) die_datenew();
  if (substdio_put(&ssout,bounce->s,bounce->len) == -1) die_datenew();
  if (substdio_flush(&ssout) == -1) die_datenew();
  if (fsync(fd) == -1) die_datenew();
  if (close(fd) == -1) die_datenew(); /* NFS stupidity */
  if (rename(fndatenew.s,fndate.s) == -1)
    strerr_die6sys(111,FATAL,ERR_MOVE,fndatenew.s," to ",fndate.s,": ");
}

void main(argc,argv)
int argc;
char **argv;
{
  char *local;
  char *host;
  char *action;
  char *def;
  int flagdig = 1;
  int flaghaveintro;
  int flaghaveheader;
  int match;
  unsigned long msgnum;
  unsigned int i;
  unsigned int len;
  char *cp;

  umask(022);
  sig_pipeignore();

  when = (unsigned long) now();

  dir = argv[1];
  if (!dir) die_usage();
  if (*dir == '-') {
    if (dir[1] == 'd') {
      flagdig = 2;
    } else if (dir[1] == 'D') {
      flagdig = 0;
    } else
      die_usage();
    dir = argv[2];
    if (!dir) die_usage();
  }
  if (chdir(dir) == -1)
    strerr_die4sys(111,FATAL,ERR_SWITCH,dir,": ");

  sender = env_get("SENDER");
  def = env_get("DEFAULT");
  local = env_get("LOCAL");

  getconf_line(&outhost,"outhost",1,FATAL,dir);
  getconf_line(&outlocal,"outlocal",1,FATAL,dir);
  workdir = dir;
  if (def) {				/* qmail>=1.02 */
    action = def;			/* now see if -digest-return- */
    if (flagdig == 1) {
      flagdig = 0;
      if (str_len(local) >= str_len(def) + 14)
	if (str_start(local + str_len(local) - 14 - str_len(def),"digest-"))
	  flagdig = 2;
    }
  } else {				/* older version of qmail */
    getconf_line(&inlocal,"inlocal",1,FATAL,dir);
    if (inlocal.len > str_len(local)) die_badaddr();
    if (case_diffb(inlocal.s,inlocal.len,local)) die_badaddr();
    action = local + inlocal.len;
    if (flagdig == 1) {
      flagdig = 0;
      if (str_start(action,"-digest")) {
	flagdig = 2;
	action += 7;
      }
    }
    if (!str_start(action,"-return-")) die_badaddr();
    action += 8;
  }
  if (flagdig) {
    if (!stralloc_copys(&ddir,dir)) die_nomem();
    if (!stralloc_cats(&ddir,"/digest")) die_nomem();
    if (!stralloc_0(&ddir)) die_nomem();
    workdir = ddir.s;
    if (!stralloc_cats(&outlocal,"-digest")) die_nomem();
  }
  if (!*action) die_trash();

  substdio_fdbuf(&ssin,read,0,inbuf,sizeof(inbuf));

  if (!case_diffs(action,"receipt")) {
    host = sender + str_rchr(sender,'@');
    if (*host)
      *(host++) = '\0';
    cp = sender;
				/* check recipient in case it's a bounce*/
    while (*(cp++)) {		/* decode sender */
      cp += str_chr(cp,'-');
      if (case_starts(cp,"-return-")) {
        if (!scan_ulong(cp + 8,&msgnum))
          strerr_die2x(100,FATAL,"bad VERP format for receipt");
         *cp = '\0';
        if (!stralloc_copys(&listaddr,sender)) die_nomem();
        if (!stralloc_append(&listaddr,"@")) die_nomem();
        if (!stralloc_cats(&listaddr,host)) die_nomem();
        if (!stralloc_0(&listaddr)) die_nomem();
        break;
      }
    }
    for(;;) {						/* Get X-tag from hdr*/
      if (getln(&ssin,&line,&match,'\n') == -1) die_msgin();
      if (!match)
        break;

      if (line.len == 1) break;
      if (case_startb(line.s,line.len,TXT_TAG)) {
	len = str_len(TXT_TAG);
        if (!stralloc_catb(&tagline,line.s +len,line.len - len -1)) die_nomem();
		/* NOTE: tagline is dirty! We quote it for sql and rely on */
		/* std log clean for maillog */
        break;
      }
    }
		/* feedback ok even if not sub. Will be filtered by subreceipt*/
		/* For instance, main list feedback is ok, but !issub. */
    subreceipt(workdir,msgnum,&tagline,listaddr.s,2,INFO,FATAL);
    closesql();
    _exit(0);
  }
				/* not receipt - maybe bounce */
				/* no need to lock. dttt.pid can be assumed */
				/* to be unique and if not would be over- */
				/* written even with lock */
  action += scan_ulong(action,&msgnum);
  if (*action != '-') die_badaddr();
  ++action;
		/* scan bounce for tag. It'll be in the BODY! */
  for (;;) {
    if (getln(&ssin,&line,&match,'\n') == -1)
      strerr_die2sys(111,FATAL,ERR_READ_INPUT);
    if (!match) break;
    if (case_startb(line.s,line.len,TXT_TAG)) {
      len = str_len(TXT_TAG);
      if (!stralloc_catb(&tagline,line.s +len,line.len - len -1)) die_nomem();
		/* NOTE: tagline is dirty! We quote it for sql and rely on */
		/* std log clean for maillog */
      break;
    }
  }
  if (seek_begin(0) == -1)
    strerr_die2sys(111,FATAL,ERR_SEEK_INPUT);
  substdio_fdbuf(&ssin,read,0,inbuf,sizeof(inbuf));

  if (*action) {	/* normal bounce */

    if (slurpclose(0,&bounce,1024) == -1) die_msgin();
    i = str_rchr(action,'=');
    if (!stralloc_copyb(&listaddr,action,i)) die_nomem();
    if (action[i]) {
      if (!stralloc_cats(&listaddr,"@")) die_nomem();
      if (!stralloc_cats(&listaddr,action + i + 1)) die_nomem();
    }
    if (!stralloc_0(&listaddr)) die_nomem();
		/* don't check for sub, since issub() doesn't see sublists */
    switch (subreceipt(workdir,msgnum,&tagline,listaddr.s,-1,INFO,FATAL)) {
	case -1: strerr_die2x(0,INFO,ERR_COOKIE);
	case -2: strerr_die2x(0,INFO,ERR_NOT_ACTIVE);
	default: doit(listaddr.s,msgnum,when,&bounce);
    }
    closesql();
    _exit(0);
  }			/* pre-VERP bounce, in QSBMF format */

  flaghaveheader = 0;
  flaghaveintro = 0;

  for (;;) {
    if (!stralloc_copys(&paragraph,"")) die_nomem();
    for (;;) {
      if (getln(&ssin,&line,&match,'\n') == -1)
        strerr_die2sys(111,FATAL,ERR_READ_INPUT);
      if (!match) die_trash();
      if (!stralloc_cat(&paragraph,&line)) die_nomem();
      if (line.len <= 1) break;
    }

    if (!flaghaveheader) {
      if (!stralloc_copy(&header,&paragraph)) die_nomem();
      flaghaveheader = 1;
      continue;
    }

    if (!flaghaveintro) {
      if (paragraph.s[0] == '-' && paragraph.s[1] == '-')
        continue;		/* skip MIME boundary if it exists */
      if (paragraph.len < 15) die_trash();
      if (str_diffn(paragraph.s,"Hi. This is the",15)) die_trash();
      if (!stralloc_copy(&intro,&paragraph)) die_nomem();
      flaghaveintro = 1;
      continue;
    }

    if (paragraph.s[0] == '-')
      break;

    if (paragraph.s[0] == '<') {	/* find address */
      if (!stralloc_copy(&failure,&paragraph)) die_nomem();

      if (!stralloc_copy(&bounce,&header)) die_nomem();
      if (!stralloc_cat(&bounce,&intro)) die_nomem();
      if (!stralloc_cat(&bounce,&failure)) die_nomem();

      i = byte_chr(failure.s,failure.len,'\n');
      if (i < 3) die_trash();

      if (!stralloc_copyb(&listaddr,failure.s + 1,i - 3)) die_nomem();
      if (byte_chr(listaddr.s,listaddr.len,'\0') == listaddr.len) {
        if (!stralloc_0(&listaddr)) die_nomem();
        if (subreceipt(workdir,msgnum,&tagline,listaddr.s,-1,INFO,FATAL) == 0)
	  doit(listaddr.s,msgnum,when,&bounce);
      }
    }
  }
  closesql();
  _exit(0);
}

